zilliqa/src/exec.rs

```diff
@@ -826,14 +890,183 @@ impl PendingState {
         }
     }
 
-    pub fn load_account(&mut self, address: Address) -> Result<&mut Account> {
-        match self.new_state.entry(address) {
-            Entry::Occupied(entry) => Ok(entry.into_mut()),
-            Entry::Vacant(vac) => {
-                let account = self.pre_state.get_account(address)?;
-                Ok(vac.insert(account))
+    pub fn load_account(&mut self, address: Address) -> Result<&mut PendingAccount> {
+        load_account(&self.pre_state, &mut self.new_state, address)
+    }
+
+    pub fn load_var_info(&mut self, address: Address, variable: &str) -> Result<(&str, u8)> {
+        let account = self.load_account(address)?;
+        let Code::Scilla { types, .. } = &account.account.code else {
+            return Err(anyhow!("not a scilla contract"));
+        };
+        let (ty, depth) = types
+            .get(variable)
+            .ok_or_else(|| anyhow!("missing type for variable: {variable}"))?;
+        Ok((ty, *depth))
+    }
+
+    /// Set the value of a given variable at the given indices. This can be used for setting map values, unlike
+    /// `load_storage`.
+    pub fn set_storage(
+        &mut self,
+        address: Address,
+        var_name: &str,
+        indices: &[Vec<u8>],
+        value: StorageValue,
+    ) -> Result<()> {
+        let account = load_account(&self.pre_state, &mut self.new_state, address)?;
+
+        let mut current = account
+            .storage
+            .entry(var_name.to_owned())
+            .or_insert_with(StorageValue::incomplete);
+        for key in indices {
+            let current_map = match current {
+                StorageValue::Map { map, .. } => map,
+                StorageValue::Value(Some(_)) => {
+                    return Err(anyhow!("expected a map"));
+                }
+                StorageValue::Value(None) => {
+                    unimplemented!("deletes of whole maps are unsupported")
```

Comment:
It's hard to tell  but I think ZQ1 supports this case -  needs an issue or a fix?
At this point you will also need a StorageValue::Deleted ...

Reply:
Just for context - Both the `unimplemented!`s are related to the assertion here (which you also commented on):
https://github.com/Zilliqa/zq2/blob/e660f240c5becf704a4af548589649878c3e7253/zilliqa/src/scilla.rs#L738

In the sense that, if that assertion holds, these branches are unreachable. I realise now that I was meant to use `unreachable!` instead.

In terms of whether we do actually need to support map-level deletions, I did some digging and couldn't find Scilla syntax for doing so, nor could I find any contracts which fail my assertion. I figured if we did find such a contract later (e.g. when replaying ZQ1 mainnet), the assertion would explode violently and we could do the necessary work to fix it?